### ISSUE TITLE
YOLO26 - how to use end2end mode with export onnx model

### ISSUE THREAD

**Issue by hexianbin1994** [ultralytics/ultralytics#24697]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I check the ultralytics code ,when use pt model , load model has end2end parameter and model.end2end is True, it will use simple nms; but when export it to onnx model, this model has no end2end parameter and model.end2end is False,  it will use normal nms and cost extra time, how to export onnx model and use end2end mode.

export onnx model:
```python
from ultralytics import YOLO
model = YOLO("best.pt")
model.export(format="onnx",opset=13,end2end=True)
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @hexianbin1994, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue routed quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this appears to be a usage/export ❓ question, please provide as much information as possible so the team can review it efficiently, including:
- your `ultralytics` version
- your Python, PyTorch, and ONNX runtime versions
- the exact export command used
- whether the behavior differs between the `.pt` model and exported `.onnx` model
- any relevant logs or console output
- a small reproducible example if possible 🙏

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by hexianbin1994** [comment]
env version info:
ultralytics=8.4.56
python=3.8
pytorch=2.1.1
onnx=1.12.0

**Comment by raimbekovm** [comment]
your export command is already correct, and onnx export keeps the end2end branch by default for yolo26, so you don't need to do anything special. the easiest way to confirm is the output shape:

- `(1, 300, 6)` means end2end is on: the graph already returns the top-300 boxes as `[x1, y1, x2, y2, conf, cls]`, no external nms needed
- `(1, 84, 8400)` means end2end is off: raw head output that still needs nms

i exported `yolo26n.pt` here to check both paths:

```python
from ultralytics import YOLO
YOLO("yolo26n.pt").export(format="onnx", opset=13)                  # output (1, 300, 6)  end2end
YOLO("yolo26n.pt").export(format="onnx", opset=13, end2end=True)    # output (1, 300, 6)  end2end
YOLO("yolo26n.pt").export(format="onnx", opset=13, end2end=False)   # output (1, 84, 8400) needs nms
```

the end2end graph uses a topk op and contains no nms node, and the export writes `end2end=True` into the onnx metadata. so when you run the exported model through `predict`/`val`, it only applies a confidence filter rather than real nms, and the detections match the `.pt` model. there's no extra nms cost in that path.

so on 8.4.56 your `end2end=True` already produces an end2end onnx. if you still get the `(1, 84, 8400)` shape, then `best.pt` isn't an end2end checkpoint (a plain yolov8/yolo11 detect model has no one2one head), and only yolo26/yolov10 weights expose the end2end branch. see the [export docs](https://docs.ultralytics.com/modes/export/) and the [yolo26 page](https://docs.ultralytics.com/models/yolo26/) for the format details.

**Comment by hexianbin1994** [comment]
> your export command is already correct, and onnx export keeps the end2end branch by default for yolo26, so you don't need to do anything special. the easiest way to confirm is the output shape:
> 
> * `(1, 300, 6)` means end2end is on: the graph already returns the top-300 boxes as `[x1, y1, x2, y2, conf, cls]`, no external nms needed
> * `(1, 84, 8400)` means end2end is off: raw head output that still needs nms
> 
> i exported `yolo26n.pt` here to check both paths:
> 
> from ultralytics import YOLO
> YOLO("yolo26n.pt").export(format="onnx", opset=13)                  # output (1, 300, 6)  end2end
> YOLO("yolo26n.pt").export(format="onnx", opset=13, end2end=True)    # output (1, 300, 6)  end2end
> YOLO("yolo26n.pt").export(format="onnx", opset=13, end2end=False)   # output (1, 84, 8400) needs nms
> the end2end graph uses a topk op and contains no nms node, and the export writes `end2end=True` into the onnx metadata. so when you run the exported model through `predict`/`val`, it only applies a confidence filter rather than real nms, and the detections match the `.pt` model. there's no extra nms cost in that path.
> 
> so on 8.4.56 your `end2end=True` already produces an end2end onnx. if you still get the `(1, 84, 8400)` shape, then `best.pt` isn't an end2end checkpoint (a plain yolov8/yolo11 detect model has no one2one head), and only yolo26/yolov10 weights expose the end2end branch. see the [export docs](https://docs.ultralytics.com/modes/export/) and the [yolo26 page](https://docs.ultralytics.com/models/yolo26/) for the format details.

ok,thanks a lot , i reinstall ultralytics and resolve it ,maybe i installed an uncorrect ultralytics

**Comment by glenn-jocher** [comment]
Glad it’s resolved — that was likely a mixed installation or shadowed package, so if it comes back please check the active Ultralytics version and import path first.

**Comment by hexianbin1994** [comment]
> > your export command is already correct, and onnx export keeps the end2end branch by default for yolo26, so you don't need to do anything special. the easiest way to confirm is the output shape:
> > 
> > * `(1, 300, 6)` means end2end is on: the graph already returns the top-300 boxes as `[x1, y1, x2, y2, conf, cls]`, no external nms needed
> > * `(1, 84, 8400)` means end2end is off: raw head output that still needs nms
> > 
> > i exported `yolo26n.pt` here to check both paths:
> > from ultralytics import YOLO
> > YOLO("yolo26n.pt").export(format="onnx", opset=13)                  # output (1, 300, 6)  end2end
> > YOLO("yolo26n.pt").export(format="onnx", opset=13, end2end=True)    # output (1, 300, 6)  end2end
> > YOLO("yolo26n.pt").export(format="onnx", opset=13, end2end=False)   # output (1, 84, 8400) needs nms
> > the end2end graph uses a topk op and contains no nms node, and the export writes `end2end=True` into the onnx metadata. so when you run the exported model through `predict`/`val`, it only applies a confidence filter rather than real nms, and the detections match the `.pt` model. there's no extra nms cost in that path.
> > so on 8.4.56 your `end2end=True` already produces an end2end onnx. if you still get the `(1, 84, 8400)` shape, then `best.pt` isn't an end2end checkpoint (a plain yolov8/yolo11 detect model has no one2one head), and only yolo26/yolov10 weights expose the end2end branch. see the [export docs](https://docs.ultralytics.com/modes/export/) and the [yolo26 page](https://docs.ultralytics.com/models/yolo26/) for the format details.
> 
> ok,thanks a lot , i reinstall ultralytics and resolve it ,maybe i installed an uncorrect ultralytics

but it looks like cause error python version, i test two case:
1 -  python3.8 and torch==2.2.2, when load model with `model=torch.load("best.pt")`, this model has no attr end2end ,model out likes yolov11 mode;
2 - python 3.10 and torch==2.2.2, same load model with `model=torch.load("best.pt")`, this model has attr end2end ,and model out is yolo26 mode, output has tensor,{one2many,scores,feats} and tensor shape is [1,300,6]

**Comment by raimbekovm** [comment]
`end2end` isn't stored in the checkpoint, so the python version itself doesn't change it. it's a computed property that reads from the installed ultralytics code at access time:

```python
# Detect.end2end
return getattr(self, "_end2end", True) and hasattr(self, "one2one")
```

what the `.pt` actually stores is the head data (`one2one_cv2` / `one2one_cv3` submodules). whether the model reports `end2end=True` is then decided by the `end2end`/`one2one` properties that ship with whatever ultralytics version is active in each env. these properties were added with the yolo26 release, so an older build resolves `end2end` differently for the same file.

so this is almost certainly an environment difference, not a python 3.8 vs 3.10 one. could you run this in both envs and paste the output:

```python
import ultralytics, torch
print(ultralytics.__version__)
m = torch.load("best.pt", weights_only=False)["model"]
print(type(m.model[-1]).__name__, getattr(m, "end2end", "NO ATTR"))
```

also, don't load with raw `torch.load`. use `YOLO("best.pt")` or `attempt_load_weights`, which run the backward-compat remapping. raw `torch.load` rebuilds objects purely from the installed classes and skips that, which is where these mismatches come from.
